sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoin.scala

```diff
@@ -61,6 +61,7 @@ case class SortMergeJoin(
     keys.map(SortOrder(_, Ascending))
 
   protected override def doExecute(): RDD[InternalRow] = {
+    // TODO(josh): why is this copying necessary?
     val leftResults = left.execute().map(_.copy())
```

Comment:
I noticed that SortMergeJoin has this defensive copying on both inputs.  I think that this is overly-conservative: we should only need to copy UnsafeRows rows that might be buffered and we should be able to perform that copying at the last possible moment when inserting the rows into the buffers.  This means that the stream side of a left or right outer join should not need to be copied.